### PR TITLE
Astar

### PR DESCRIPTION
**Pull request by asymetrique**
A* path planning infrastructure added, however it is not used, yet.
Dog's path is represented as a continuous Bezier curve

### CODE REVIEW

File: levelmap.go [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+		image.Pt(0, 1),  // South
+		image.Pt(-1, 0), // West
+	}
+	offsetsDiag := []image.Point{
+		image.Pt(1, -1),  // NorthEast
+		image.Pt(1, 1),   // SouthEast
+		image.Pt(-1, 1),  // SouthWest
+		image.Pt(-1, -1), // NorthWest
+	}
+	res := make([]image.Point, 0, 8)
```
Comment (sinisterstuf) [inline levelmap.go:40]: These things look actually like `consts` that exist only to be referenced later (e.g. by name) but are not expected to vary later, am I right?

File: levelmap.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	lmap := make([][]int, h)
+	for i := range lmap {
+		lmap[i] = make([]int, w)
+	}
+	
+	return lmap
+}
+
+// Neighbours implements the astar.Graph interface
+func (m LevelMap) Neighbours(p image.Point) []image.Point {
```
Comment (sinisterstuf) [inline levelmap.go:27]: It's unusual to see in code but :heart: for the British spelling 🇬🇧

File: levelmap.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+// FindPath finds path between two coordinates on map
+func (m LevelMap) FindPath(start, dest Coord) []Coord {
+	var result []Coord
+
+	startCell := image.Pt(int(start.X / 32), int(start.Y / 32))
+	destCell := image.Pt(int(dest.X / 32), int(dest.Y / 32))
+	apath := astar.FindPath[image.Point](m, startCell, destCell, distance, distance)
+	apath = simplifyPath(apath)
+	for _, p := range apath {
+		result = append(result, Coord{X: float64(p.X)*32+16, Y: float64(p.Y)*32+16})
```
Comment (sinisterstuf) [inline levelmap.go:87]: I'm amazed that you managed to mix image.Point and float64 in the same file and hardly had to do any manual type casts. Not sure if lucky or very clever. Btw these hard-coded `32` and `16` values are probably referring to some size thing that's depending on something from some external resource I guess, tiles / map data or something like that right?

File: levelmap.go [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+	return res
+}
+
+// isFreeAt returns if the cell is free
+func (m LevelMap) isFreeAt(p image.Point) bool {
+	return m[p.Y][p.X] == 0
+}
+
+// distance calculates Euclidean distance between the points
+func distance(p, q image.Point) float64 {
```
Comment (sinisterstuf) [inline levelmap.go:68]: Hmmm this looks almost exactly the same as another utility function we have lying in main.go, the usage is even almost the same, I guess it's just too bad that the types differ, this one is using image.Point and the other is using float64. Oh well.

File: levelmap.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+	offsetsDiag := []image.Point{
+		image.Pt(1, -1),  // NorthEast
+		image.Pt(1, 1),   // SouthEast
+		image.Pt(-1, 1),  // SouthWest
+		image.Pt(-1, -1), // NorthWest
+	}
+	res := make([]image.Point, 0, 8)
+	
+	// Check avaialable diagonal neighbours (free only if there is a wide enough corridor)
+	for _, off := range offsetsDiag {
```
Comment (sinisterstuf) [inline levelmap.go:43]: This is just a nitpick, but naming this variable `off` completely threw me off because I thought it was going to represent some boolean state, like is it on/off, but it's actually short for offset. Maybe `offset` or `o` would avoid this ambiguity?

File: levelmap.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+		q := p.Add(off)
+		q1 := p.Add(image.Pt(0, off.Y))
+		q2 := p.Add(image.Pt(off.X, 0))
+		if m.isFreeAt(q) && m.isFreeAt(q1) && m.isFreeAt(q2) {
```
Comment (sinisterstuf) [inline levelmap.go:47]: I'm guessing this is necessary because if you're not moving just up/down/left/right but also diagonally then because the diagonal tiles hardly touch, they wouldn't be considered, but they should be, am I right?

File: levelmap.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+	}
+
+	// Check avaialable  neighbours
+	for _, off := range offsets {
+		q := p.Add(off)
+		if m.isFreeAt(q) {
+			res = append(res, q)
+		}
+	}
+	return res
```
Comment (sinisterstuf) [inline levelmap.go:59]: `res` :thinking: this looks suspiciously like JavaScript :sweat_smile: if it's possible to not have every function using variables like `this` or `res` but rather names that describe what the things are or what we do with them that'd be cool, like in this case the function is about Neighbours, so we I guess it's returning a list of available neighbouring cells, so maybe… `neighbour`? :thinking:

File: levelmap.go [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+func (m LevelMap) SetObstacle(x, y int) {
+	m[y][x] = 1
+}
+
+// FindPath finds path between two coordinates on map
+func (m LevelMap) FindPath(start, dest Coord) []Coord {
+	var result []Coord
+
+	startCell := image.Pt(int(start.X / 32), int(start.Y / 32))
+	destCell := image.Pt(int(dest.X / 32), int(dest.Y / 32))
```
Comment (sinisterstuf) [inline levelmap.go:83]: What is 32? :thinking:

File: levelmap.go [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+	apath := astar.FindPath[image.Point](m, startCell, destCell, distance, distance)
+	apath = simplifyPath(apath)
```
Comment (sinisterstuf) [inline levelmap.go:85]: Why is this necessary btw? Is there a shortcoming in the pathfinding library we're using? :thinking:

**Review by sinisterstuf — COMMENTED** [review]
This is a bit much for me to take in at once, I'm not gonna lie :stuck_out_tongue_closed_eyes: so most of my comments are actually questions trying to grok what's going on.

File: main.go [hunk 10]
```diff
@@ -32,7 +37,7 @@ const (
 )
 
 func main() {
-	gameWidth, gameHeight := 320, 240
+	gameWidth, gameHeight := 640, 480
```
[the remark this answers is not in the document]
Comment (asymetrique) [reply main.go:40]: Yes, I will not merge this

File: main.go [hunk 11]
```diff
@@ -20,6 +20,11 @@ import (
 	camera "github.com/melonfunction/ebiten-camera"
 	"github.com/solarlune/ldtkgo"
 	"github.com/solarlune/resolv"
+
+	beziercp "github.com/brothertoad/bezier"
+	beziercurve "gonum.org/v1/plot/tools/bezier"
+	"gonum.org/v1/plot/vg"
+	"gonum.org/v1/plot/font"
```
[the remark this answers is not in the document]
Comment (asymetrique) [reply main.go:27]: The unit used by vg lib is font.Length. I hate it, but don't know how to overcome. It does not accept anything else, however in the end it's just float64

File: main.go [hunk 12]
```diff
@@ -195,14 +205,30 @@ func NewGame(g *Game) {
...
+	var dogpath []Coord
+	for _, c := range curveCPs {
+		curve := beziercurve.New(
+			vg.Point{X: font.Length(int(c.P0.X)), Y: font.Length(c.P0.Y)},
+			vg.Point{X: font.Length(int(c.P1.X)), Y: font.Length(c.P1.Y)},
+			vg.Point{X: font.Length(int(c.P2.X)), Y: font.Length(c.P2.Y)},
+			vg.Point{X: font.Length(int(c.P3.X)), Y: font.Length(c.P3.Y)},
+		)
+
+		// 4 points per curve
```
[the remark this answers is not in the document]
Comment (asymetrique) [reply main.go:228]: I didn't want to create too many intermediate points, and the expereince seemed smooth to me. We can increase/finetune it to have even smoother path.

File: main.go [hunk 13]
```diff
@@ -133,6 +141,8 @@ func NewGame(g *Game) {
 					float64(layer.GridSize),
 					tagWall,
 				))
+
+				g.LevelMap.SetObstacle(intData.Position[0]/32, intData.Position[1]/32)
```
[the remark this answers is not in the document]
Comment (asymetrique) [reply main.go:145]: Yes, sure :) I will do

File: levelmap.go (hunk 1 above)
Comment (asymetrique) [reply levelmap.go:40]: Yes. What to do with it then?

File: levelmap.go (hunk 5 above)
Comment (asymetrique) [reply levelmap.go:43]: OK

File: levelmap.go (hunk 6 above)
Comment (asymetrique) [reply levelmap.go:47]: Exactly. I would allow diagonal movement only if there is enough space

File: levelmap.go (hunk 7 above)
Comment (asymetrique) [reply levelmap.go:59]: OK

File: levelmap.go (hunk 8 above)
Comment (asymetrique) [reply levelmap.go:83]: It is to find out the cell/tile from the exact position

File: levelmap.go (hunk 9 above)
Comment (asymetrique) [reply levelmap.go:85]: Which part do you mean? simplify? It is because otherwise we got all the tiles along the path, but if you don't need to turn for a while (e.g just walking horizontal), then we don't need all the tiles, just the end tile. It decreases the tile number of the path significantly.

File: levelmap.go (hunk 4 above)
Comment (asymetrique) [reply levelmap.go:68]: Yes. We may also be able to use image.Point internally, instead of our own similar struct ??

File: levelmap.go (hunk 2 above)
Comment (asymetrique) [reply levelmap.go:27]: :) I borrowed this code

File: levelmap.go (hunk 3 above)
Comment (asymetrique) [reply levelmap.go:87]: I think I will create a Level object that can contain all of these 16, 32, ... stuff from ldtk and it can also contain the map for pathfinding then.

Comment (asymetrique) [reply levelmap.go:87]: No, I change my mind... I think I will just have a constant for this for the time being and introduce Level if we will have more levels. I think Level handling would be the right solution, but I would not like to introduce further unnecessary complexity.

File: levelmap.go (hunk 9 above)
Comment (sinisterstuf) [reply levelmap.go:85]: Ah ok, makes so much more sense now haha, thanks!

File: levelmap.go (hunk 3 above)
Comment (sinisterstuf) [reply levelmap.go:87]: Ah, I'm happy I found this comment again and I'm glad you changed your mind cos I was going to ask you not to do this

> I think I will create a Level object that can contain all of these 16, 32, ... stuff from ldtk

What advantage would this bring? Just a shorter name to call it by? These values are already provided by ldtk-go we just need to use them. For example, if the 32 here is the layer's grid size, that's found in [layer.GridSize](https://pkg.go.dev/github.com/solarlune/ldtkgo#Layer). A usage example from our own code: [961ae68/main.go#L133-L134](https://github.com/sinisterstuf/escort-mission/blob/961ae684345066f358bc63b99f0fc14172fbfa56/main.go#L133-L134). The 16 here I'm not sure what it represents (because it has no name) but I suppose it's GridSize/2 to be able to get the centre of it, but it may also be some other value from the map data, in which case I'm pretty sure we can find it from the ldtk-go-provided data 😉

File: levelmap.go (hunk 8 above)
Comment (sinisterstuf) [reply levelmap.go:83]: ok, maybe you could use [Layer.ToGridPosition()](https://pkg.go.dev/github.com/solarlune/ldtkgo#Layer.ToGridPosition) for this? it takes coordinates as argument and returns a cell index and it is aware of the grid's cell size 16/32/64 whatever so you don't have to mention it

File: levelmap.go (hunk 4 above)
Comment (sinisterstuf) [reply levelmap.go:68]: I don't recommend this unless you want to make use of lots of these Point-specific functions that are already implemented in the standard library, add, subtract, and checking whether it's within another rectangle etc. otherwise it's annoying. I did something similar in a previous game, I thought it would be so cute and clean if the whole game was modeled around the stdlib's image.Point and image.Rect etc. but then 50% of my code was just converting between `int` and `float64` 😭

File: levelmap.go (hunk 1 above)
Comment (sinisterstuf) [reply levelmap.go:40]: Nothing I guess 😂 you can't make array constants, life is life.

OTOH you could give them all names if you like:

```go
const N = something
const NE = idk
cost E = ...etc
```

no idea...

File: levelmap.go (hunk 4 above)
Comment (asymetrique) [reply levelmap.go:68]: OK, thx
